man/man3/game_assert.3
.\" Copyright (c) 2025 Mattia Cabrini
.\" SPDX-License-Identifier: GFDL-1.3-only

.TH game_assert 3 2025-09-22 "CMC-Chess Source Code Man-Pages pre-release"

.SH NAME
game_assert.h, game_assert.c, struct game_assert_t, game_assert_p \- Parse an assert command

.SH SYNOPSIS

.Bd
\fB#include "game_assert.h"\fP
.Ed
.P
.EX
enum
{
    ASSERT_KIND_UNKNOWN, 

    ASSERT_KIND_CHECK, 
    ASSERT_KIND_CHECKMATE, 
    ASSERT_KIND_PIECE_IS, 
    ASSERT_KIND_PIECE_CAN_MOVE
};
.P
typedef struct game_assert_t
{
    int kind;
    int rev;

    piece_t piece;
    turn_t  turn;

    struct coord_t src;
    struct coord_t dst;
    struct coord_t whence;
}* game_assert_p;
.P
.B void game_assert_parse(game_assert_p A, const char* str, char* err, size_t err_length);

.P
.B Private
static const char* parse_kind_string(game_assert_p A, const char* str);
static const char* parse_int(int* n, const char* str);
static const char* parse_attribute_name(const char* str, char attr_name, size_t attr_name_n);
static const char* parse_int(int* n, const char* str);
.P
#define MAX_ATTR_NAME_LENGTH 16

.SH DESCRIPTION

.P
.BR "game_assert_parse()"
attempts to parse
.I str
(an assert command described in 
.BR game_assert(5)
). In case of error, 
.I  "A->kind"
is set to ASSERT_KIND_UNKNOWN and an error message is copied into
.I err
up to a maximum of
.I err_length
characters.

.P
.BR "struct game_assert_t"
.TP
.B kind
The kind of assert (see enum below): it tells what kind of check shall be performed, and therefore how other members shall be used.
.TP
.B rev
If set to false the assertion MUST have positive match; If set to true the assertions is intended as 
.I  "reversed"
 and therefore MUST turn out as negative match.
.TP
.B piece
In case of
.B ASSERT_KIND_PIECE_IS
.I piece
indicates what must be matched in position
.I src
and it is ignored otherwise.
.TP
.B turn
Override current turn. A positive value indicate to assume that it is WHITE turn; A negative value indicates to assume that it is BLACK turn; 0 is undefined behaviour.
.TP
.B src
Source coordinates.
.TP
.B dst
Destination coordinates.
.TP
.B whence
In case of ASSERT_KIND_CHECK tells whence the king should be taken over; In case of ASSERT_KIND_PIECE_CAN_MOVE tells whence the kind should be taken over, hence implying the move must not be viable.

.P
.BR "The enum contains all types of asserts."
.TP
.B ASSERT_KIND_UNKNOWN
A special kind that represent an error occurred at parse time.
.TP
.B ASSERT_KIND_CHECK
To assert that a piece in under check. This assertion requires to specify:
.I src
to tell the King position;
.I whence
to tell whence the king shall be taken over. Note that if the king is under check by multiple pieces
.I whence
behaviour is undefined and therefore
.I whence
should not be specified.
.TP
.B ASSERT_KIND_CHECKMATE
To assert that a King is under checkmate. This assertion requires to specify
.I src
to tell the position of the King.
.TP
.B ASSERT_KING_PIECE_IS
To assert that a specific piece must be placed at certain coordinates. This assertion requires to specify:
.I src
to tell what cell to look at; and
.I piece
to tell what piece is expected (see piece.3).

.SH SEE ALSO

.B cmc-chess(6), game_assert(5), piece(3)


game_assert.c
/* Copyright (c) 2025 Mattia Cabrini      */
/* SPDX-License-Identifier: AGPL-3.0-only */

#include "game_assert.h"
#include "board.h"
#include "int.h"
#include "util.h"

#include <stdlib.h>
#include <string.h>

#define MAX_ATTR_NAME_LENGTH 16

/* Read Assert Kind
 * - Move str to first not blank character;
 * - Checks str against all possible values and set A->kind.
 *
 * If all goes well return a pointer to the next char to parse.
 * An error condition can be detected by checking A->kind against
 * ASSERT_KIND_UNKNOWN.
 */
static const char* parse_kind_string(game_assert_p A, const char* str);

/* Read Attribute Name
 * - Move str to first not blank character;
 * - Read the attribute name. That is a contiguous block of not blank characters
 *   of maximum size MAX_ATTR_NAME_LENGTH (including NUL terminator). Parsing
 *   goes on until it reaches '=';
 * - If the assert kind as string is longer than MAX_ATTR_NAME_LENGTH, return
 *   NULL for it is an error condition.
 *
 * If all goes well return a pointer to the next char to parse.
 */
static const char*
parse_attribute_name(const char* str, char* attr_name, size_t attr_name_n);

/* Read An Integer
 * - Move str to the first not blank character;
 * - Read a piece, turn or a boolean. That is at most two characters in the
 *   form `[-]\d`:
 *   - If the first character is '-' the parser reads two characters.
 *   - Otherwise it reads one character.
 * - Return NULL if the first char is NUL;
 * - Return NULL if the first char is '-' and the second character is NUL;
 *
 * If all goes well return a pointer to the next char to parse.
 */
static const char* parse_int(int* n, const char* str);

/* Read Coordidates
 * - Move str to first not blank character;
 * - Read the coordinates. That is exactly two characters `[A-G][1-8]`.
 * - Return NULL if there are not two characters available;
 * - Does not check if read coordinates does not actually translate into a board
 *   cell.
 *
 * If all goes well return a pointer to the next char to parse.
 */
static const char* parse_coord(coord_p A, const char* str);

void game_assert_parse(
    game_assert_p A, const char* str, char* err, size_t err_length
)
{
    char attr_name[MAX_ATTR_NAME_LENGTH];
    int  tmp; /* For piece and turn */

    A->src.row    = -1;
    A->dst.row    = -1;
    A->whence.row = -1;
    A->pawn_morph = cpEEMPTY;

    A->rev        = 0;

    str           = parse_kind_string(A, str);
    if (A->kind == ASSERT_KIND_UNKNOWN)
    {
        strncpy(err, "could not parse assert kind", err_length);
        return;
    }

    for (; str != NULL;)
    {
        str = parse_attribute_name(str, attr_name, sizeof(attr_name));
        if (str == NULL)
        {
            strncpy(err, "could not read attribute name", err_length);
            break;
        }

        if (!*str)
            break;

        if (streq_ci(attr_name, "src"))
        {
            str = parse_coord(&A->src, str);
            if (str == NULL)
                strncpy(err, "could not read source", err_length);
        }
        else if (streq_ci(attr_name, "dst"))
        {
            str = parse_coord(&A->dst, str);
            if (str == NULL)
                strncpy(err, "could not read destination", err_length);
        }
        else if (streq_ci(attr_name, "whence"))
        {
            str = parse_coord(&A->whence, str);
            if (str == NULL)
                strncpy(err, "could not read destination", err_length);
        }
        else if (streq_ci(attr_name, "pawn-morph"))
        {
            str = parse_int(&tmp, str);
            if (str == NULL)
                strncpy(err, "could not read pawn-morph", err_length);
            else
                A->pawn_morph = (piece_t)tmp;
        }
        else if (streq_ci(attr_name, "piece"))
        {
            str = parse_int(&tmp, str);

            if (str != NULL)
            {
                A->piece = (piece_t)tmp;
                if (piece_to_char(A->piece) == 'E')
                    str = NULL;
            }

            if (str == NULL)
                strncpy(err, "could not read piece", err_length);
        }
        else if (streq_ci(attr_name, "turn"))
        {
            str = parse_int(&tmp, str);
            if (str == NULL)
                strncpy(err, "could not read turn", err_length);
            else
                A->turn = (turn_t)tmp;
        }
        else if (streq_ci(attr_name, "rev"))
        {
            str = parse_int(&tmp, str);
            if (str == NULL)
                strncpy(err, "could not read turn", err_length);
            else
                A->rev = (turn_t)tmp;
        }
    }

    if (str == NULL)
        A->kind = ASSERT_KIND_UNKNOWN;
}

static const char* parse_kind_string(game_assert_p A, const char* str)
{
    str = move_to_not_blank(str);

    if (strneq_ci(str, "piece-can-move", 14))
    {
        A->kind = ASSERT_KIND_PIECE_CAN_MOVE;
        str += 14;
    }
    else if (strneq_ci(str, "checkmate", 9))
    {
        A->kind = ASSERT_KIND_CHECKMATE;
        str += 9;
    }
    else if (strneq_ci(str, "piece-is", 8))
    {
        A->kind = ASSERT_KIND_PIECE_IS;
        str += 8;
    }
    else if (strneq_ci(str, "check", 5))
    {
        A->kind = ASSERT_KIND_CHECK;
        str += 5;
    }
    else
    {
        A->kind = ASSERT_KIND_UNKNOWN;
        return NULL;
    }

    return str;
}

static const char*
parse_attribute_name(const char* str, char* attr_name, size_t attr_name_n)
{
    str = move_to_not_blank(str);

    if (!*str)
        return str;

    for (; attr_name_n > 1 && *str && *str != '='; --attr_name_n)
    {
        *attr_name = *str;
        ++str;
        ++attr_name;
    }

    *attr_name = '\0';
    if (*str == '=')
        return ++str;

    /* Name too long */
    return NULL;
}

static const char* parse_int(int* n, const char* str)
{
    char   buf[3] = {0, 0, 0};
    size_t cur;

    cur = 0;
    str = move_to_not_blank(str);

    if (*str == '\0')
        return NULL; /* Empty */

    if (*str == '-')
        buf[cur++] = *(str++);

    buf[cur++] = *(str++);

    *n         = (piece_t)atoi(buf);

    return str;
}

static const char* parse_coord(coord_p C, const char* str)
{
    char buf[3];

    str = move_to_not_blank(str);

    if (!*str)
        return NULL; /* Err */
    buf[0] = *(str++);

    if (!*str)
        return NULL; /* Err */
    buf[1] = *(str++);
    buf[2] = '\0';

    coord_init_by_str(C, buf);

    return str;
}

#ifdef DEBUG
void game_assert_meminfo(void)
{
    struct game_assert_t T;

    printf("struct game_assert_t: %lu\n", sizeof(T));
    printf(" kind:                %lu\n", sizeof(T.kind));
    printf(" rev:                 %lu\n", sizeof(T.rev));
    printf(" src:                 %lu\n", sizeof(T.src));
    printf(" dst:                 %lu\n", sizeof(T.dst));
    printf(" whence:              %lu\n", sizeof(T.whence));
    printf(" piece:               %lu\n", sizeof(T.piece));
    printf(" turn:                %lu\n", sizeof(T.turn));
    printf(" pawn_morph:          %lu\n", sizeof(T.pawn_morph));
    printf(
        " -------------------- %lu\n",
        sizeof(T.kind) + sizeof(T.rev) + sizeof(T.piece) + sizeof(T.turn) +
            sizeof(T.src) + sizeof(T.dst) + sizeof(T.whence) +
            sizeof(T.pawn_morph)
    );
}
#endif
